src/NuGet.Core/NuGet.Packaging/Licenses/LicenseExpressionParser.cs

```diff
@@ -0,0 +1,211 @@
+        }
+
+        /// <summary>
+        /// Based on the Shunting Yard algorithm. <see href="https://en.wikipedia.org/wiki/Shunting-yard_algorithm"/>
+        /// Given an array of tokens, generates a NuGetLicenseExpression if valid.
+        /// </summary>
+        /// <param name="infixTokens"></param>
+        /// <returns>NuGetLicenseExpression</returns>
+        /// <exception cref="ArgumentException">If the passed in values are not a valid SPDX License Expression.</exception>
+        private static NuGetLicenseExpression Parse(IEnumerable<LicenseExpressionToken> infixTokens)
+        {
+            var operatorStack = new Stack<LicenseExpressionToken>();
+            var operandStack = new Stack<LicenseExpressionToken>();
+            NuGetLicenseExpression leftHandSideExpression = null;
+            NuGetLicenseExpression rightHandSideExpression = null;
+
+            var lastTokenType = LicenseTokenType.VALUE;
+            var firstPass = true;
+
+            foreach (var token in infixTokens)
+            {
+                var currentTokenType = token.TokenType;
+                switch (token.TokenType)
+                {
+                    case LicenseTokenType.OPENING_BRACKET:
+                        if (!firstPass && !token.TokenType.IsValidPrecedingToken(lastTokenType))
+                        {
+                            throw new ArgumentException(string.Format(CultureInfo.CurrentCulture, Strings.NuGetLicenseExpression_InvalidToken, token.Value));
+                        }
+                        operatorStack.Push(token);
+                        break;
+
+                    case LicenseTokenType.CLOSING_BRACKET:
+                        if (firstPass || !token.TokenType.IsValidPrecedingToken(lastTokenType))
+                        {
+                            throw new ArgumentException(string.Format(CultureInfo.CurrentCulture, Strings.NuGetLicenseExpression_InvalidToken, token.Value));
+                        }
+
+                        // pop until we hit the opening bracket
+                        while (operatorStack.Count > 0 && operatorStack.Peek().TokenType != LicenseTokenType.OPENING_BRACKET)
+                        {
+                            ProcessOperators(operatorStack, operandStack, ref leftHandSideExpression, ref rightHandSideExpression);
+                        }
+
+                        if (operatorStack.Count > 0)
+                        {
+                            // pop the bracket
+                            operatorStack.Pop();
+                        }
+                        else
+                        {
+                            throw new ArgumentException(string.Format(CultureInfo.CurrentCulture, Strings.NuGetLicenseExpression_MismatchedParenthesis));
+                        }
+                        break;
+
+                    case LicenseTokenType.VALUE:
+                        if (!firstPass && !token.TokenType.IsValidPrecedingToken(lastTokenType))
```

Comment:
What do you think of moving this case to the top for readability's sake? It took me a minute to realize the difference between the `operandStack` and the `operatorStack`.

Reply:
Moved it. 

The operand/operator stack should be self descriptive though? 

I will add some more comments about how the algorithm works overall and ping you.